sound/soc/sof/fw-file-profile.c

```diff
@@ -91,37 +198,110 @@ sof_file_profile_for_ipc_type(struct device *dev,
+	char *marker;
+	int i;
+
+	dev_err(dev, "SOF firmware and/or topology file not found.\n");
+	dev_info(dev, "Supported default profiles\n");
+
+	for (i = 0; i < SOF_IPC_TYPE_COUNT; i++) {
+		if (!(desc->ipc_supported_mask & BIT(i)))
+			continue;
+
+		if (i == ipc_type)
+			marker = "Requested";
+		else
+			marker = "Fallback";
+
+		dev_info(dev, "- ipc type %d (%s):\n", i, marker);
+		if (base_profile->fw_path_postfix)
+			dev_info(dev, " Firmware file: %s/%s/%s\n",
+				 desc->default_fw_path[i],
+				 base_profile->fw_path_postfix,
+				 desc->default_fw_filename[i]);
+		else
+			dev_info(dev, " Firmware file: %s/%s\n",
+				 desc->default_fw_path[i],
+				 desc->default_fw_filename[i]);
+
+		dev_info(dev, " Topology file: %s/%s\n",
+			 desc->default_tplg_path[i],
+			 plat_data->tplg_filename);
+	}
+
+	if (base_profile->fw_path || base_profile->fw_name ||
+	    base_profile->tplg_path || base_profile->tplg_name)
+		dev_info(dev, "Verify the path/name override module parameters.\n");
+
+	dev_info(dev, "Check if you have 'sof-firmware' package installed.\n");
+	dev_info(dev, "Optionally it can be manually downloaded from:\n");
+	dev_info(dev, "   https://github.com/thesofproject/sof-bin/\n");
+}
+
+static void sof_print_profile_info(struct snd_sof_dev *sdev,
+				   enum sof_ipc_type ipc_type,
 				   struct sof_loadable_file_profile *profile)
 {
+	struct device *dev = sdev->dev;
+
+	if (ipc_type != profile->ipc_type)
+		dev_info(dev,
+			 "Using fallback IPC type %d (requested type was %d)\n",
+			 profile->ipc_type, ipc_type);
+
 	dev_info(dev, "Firmware paths/files for ipc type %d:\n", profile->ipc_type);
 
 	dev_info(dev, " Firmware file:     %s/%s\n", profile->fw_path, profile->fw_name);
 	if (profile->fw_lib_path)
 		dev_info(dev, " Firmware lib path: %s\n", profile->fw_lib_path);
-	if (profile->tplg_name)
-		dev_info(dev, " Topology file:     %s/%s\n", profile->tplg_path,
-			profile->tplg_name);
-	else
-		dev_info(dev, " Topology path:     %s\n", profile->tplg_path);
+	dev_info(dev, " Topology file:     %s/%s\n", profile->tplg_path, profile->tplg_name);
 }
 
 int sof_create_ipc_file_profile(struct snd_sof_dev *sdev,
 				struct sof_loadable_file_profile *base_profile,
 				struct sof_loadable_file_profile *out_profile)
 {
 	const struct sof_dev_desc *desc = sdev->pdata->desc;
-	struct device *dev = sdev->dev;
-	int ret;
+	int ret = -ENOENT;
+	int i;
 
 	memset(out_profile, 0, sizeof(*out_profile));
 
-	ret = sof_file_profile_for_ipc_type(dev, desc, base_profile, out_profile);
-	if (ret)
-		return ret;
+	ret = sof_file_profile_for_ipc_type(sdev, base_profile->ipc_type, desc,
+					    base_profile, out_profile);
+	if (!ret)
+		goto out;
 
-	sof_print_profile_info(dev, out_profile);
+	/*
+	 * No firmware file was found for the requested IPC type, check all
+	 * IPC types as fallback
+	 */
+	for (i = 0; i < SOF_IPC_TYPE_COUNT; i++) {
+		if (i == base_profile->ipc_type ||
+		    !(desc->ipc_supported_mask & BIT(i)))
```

Comment:
> In the specific question: if for the platform the IPC3 files are missing, but the platform supports IPC4 then we will try to probe using IPC4, if that is also missing then we fail to probe.

That's a change of functionality that should be controlled with a Kconfig. For now IPC3 platforms need to work with IPC3 only, otherwise for each case where sof-bin is not properly installed we'll also have a confusing additional error with IPC4 stuff missing, or entire new problems created.

Reply:
I see, we have this print in the case you have described:
```
[  115.249648] snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw path: community
[  115.249650] snd_sof:sof_select_ipc_and_paths: sof-audio-pci-intel-tgl 0000:00:1f.3: Path postfix appended to default fw_lib path: community
[  115.249705] snd_sof:sof_test_firmware_file: sof-audio-pci-intel-tgl 0000:00:1f.3: Failed to open firmware file: intel/sof/community/sof-tgl.ri
[  115.249744] snd_sof:sof_test_firmware_file: sof-audio-pci-intel-tgl 0000:00:1f.3: Failed to open firmware file: intel/sof-ipc4/tgl/community/sof-tgl.ri
[  115.249747] sof-audio-pci-intel-tgl 0000:00:1f.3: SOF firmware and/or topology file not found.
[  115.249749] sof-audio-pci-intel-tgl 0000:00:1f.3: Supported default profiles
[  115.249751] sof-audio-pci-intel-tgl 0000:00:1f.3: - ipc type 0 (Requested):
[  115.249753] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file: intel/sof/community/sof-tgl.ri
[  115.249754] sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file: intel/sof-tplg/sof-hda-generic.tplg
[  115.249756] sof-audio-pci-intel-tgl 0000:00:1f.3: - ipc type 1 (Fallback):
[  115.249758] sof-audio-pci-intel-tgl 0000:00:1f.3:  Firmware file: intel/sof-ipc4/tgl/community/sof-tgl.ri
[  115.249760] sof-audio-pci-intel-tgl 0000:00:1f.3:  Topology file: intel/sof-ipc4-tplg/sof-hda-generic.tplg
[  115.249762] sof-audio-pci-intel-tgl 0000:00:1f.3: Check if you have 'sof-firmware' package installed.
[  115.249764] sof-audio-pci-intel-tgl 0000:00:1f.3: Optionally it can be manually downloaded from:
[  115.249766] sof-audio-pci-intel-tgl 0000:00:1f.3:    https://github.com/thesofproject/sof-bin/
```
So to prevent fallback to newer IPC version (is that a fallup?) in case the default cannot be satisfied?
CONFIG_SOF_ALLOW_FALLBACK_TO_NEWER_IPC_VERSION, and it is **n** by default?

Reply:
something like that yes.